docs/design/kusion_module/kusion_modules.md

```diff
@@ -0,0 +1,117 @@
+# Kusion Module Design Doc
+
+## Context
+
+Kusion module is a reusable building block designed by platform engineers to standardize application deployments and enable app developers to self-service. For more details, please visit our [official website](https://www.kusionstack.io/docs/kusion/concepts/kusion-module).
+
+![module](../collaboration/kusion-module.png)
+
+## Goals
+
+1. Design a flexible Kusion module mechanism to manage and use modules
+2. Provide a user-friendly framework to enable users to develop customized modules
+
+## Structure
+
+An app dev-orient schema, a generator and a license file are three components required for a legal Kusion module. We strongly recommend adding a readme file and examples in the module package for completeness. An example module package is shown as follows.
+
+```shell
+$ tree example-module/
+.
+├── schema.k
+├── generator # binary
+├── kcl.mod
+├── README.md
+├── LICENSE
+├── examples/
+│   ├── main.k
+│   ├── workspace.yaml
+```
+
+## Lifecycle
+
+### Execution lifecycle
+
+#### Download and unzip
+
+A complete set of modules of one stack consists of two parts: modules in the AppConfig and modus in the workspace. In most scenes, the two parts are the same, but modules in the workspace can be bigger than those in the AppConfig as **some modules do not contain schemas**.
+
+We need to set KPM download path as `$KUSION_HOME/modules`, since it will always download schema-related modules in a certain path to guarantee it works correctly and then Kusion will download extra modules defined in the workspace.
```

Comment:
Will this cause the existing KPM package of users need to be migrated?

Reply:
I'll discuss it with them to enable KPM to supporting multiple pkg paths.